torch/_dynamo/symbolic_convert.py

```diff
@@ -774,29 +775,129 @@ def maybe_has_backedge(self):
-                self._cell_and_freevars += self.parent.cell_and_freevars()
+                self._freevars += self.parent.freevars()
+        return self._freevars
+
+    def cell_and_freevars(self):
+        if not hasattr(self, "_cell_and_freevars"):
+            self._cell_and_freevars = self.cellvars() + self.freevars()
         return self._cell_and_freevars
 
     def prune_dead_locals(self):
         reads = livevars_analysis(self.instructions, self.current_instruction)
         # implicit use by super()
         # reads = reads | {"__class__"}
         # output variables?
-        reads = reads | set(self.cell_and_freevars())
+        reads = reads | set(self.freevars())
+
+        # First we prune the non-cell local vars, this allows us to prune more
+        # cell local vars later on (e.g., if we manage to prune a
+        # `NestedUserFunctionVariable` that makes use of some cell locals).
+        cellvars = set(self.cellvars())
         self.symbolic_locals = {
-            k: v for k, v in self.symbolic_locals.items() if k in reads
+            k: v for k, v in self.symbolic_locals.items() if k in cellvars or k in reads
         }
+
+        # Then we prune the side effects, which might enable us to prune more
+        # cellvars afterwards.
         self.output.side_effects.prune_dead_object_new(self)
 
+        # Then we prune the cell locals.
+        #
+        # Note that we keep certain cell locals, because the current mechanism
+        # for codegen closure initialization for nested function creation is:
+        # 1. `NestedUserFunctionVariable` codegen assumes its closure has been
+        #    initialized properly by its creater, i.e., the tuple of cells will
+        #    be populated with correct content before the function is used.
+        # 2. In the fast path of `OutputGraph::compile_subgraph`, we populate
+        #    the tuple of cells _after_ emitting the `MAKE_FUNCTION` bytecode,
+        #    via `STORE_DEREF`; these `STORE_DEREF` are generated partly based
+        #    on the current `symbolic_locals`.
+        # As a result, we must be careful not to prune the cell locals that'll
+        # allow `OutputGraph` to generate the proper `STORE_DEREF`.
+        #
+        # On the other hand, we do want to prune away the truly dead ones, e.g.,
+        # say after we invoke a nested function, and the function is never used
+        # again. So here we do some conservative pruning, by tracing from all
+        # living `symbolic_locals` that are _not_ cells; for any reachable cell,
+        # it must be kept alive.
+        #
+        # TODO there are extra complexities due to side-effects (e.g., the
+        # nested function leaking out into backward hook or globals). We could
+        # probably improve the variable tracing here to include the relevant
+        # variables in `output.side_effects`.
+        if self.output.side_effects.is_empty():
```

Comment:
This seems limiting. If we're worried about the nested function leaking out, would it make sense to see if any of the side effects depends on the nested function?

Reply:
IIUC this check basically says "we only attempt to prune cell variables if there is 0 side-effect". So that should guarantee no leaking? In the future we could relax this check by traversing from the variables in `SideEffects` (which could include one of those leaking `NestedFunctionVariable`). I wanted to keep my first PR a bit more innocuous:).

Reply:
I agree that that this check works, I think in practice it is too limiting and that we'll want to expand this soon. I'm happy if you're planning on taking it as a follow-up (if not, please file an issue for it so that we know we've got this problem)

Reply:
Cool I created an issue and will work on it as a follow-up: #137123.